hail/src/main/scala/is/hail/expr/ir/TableIR.scala

```diff
@@ -476,6 +476,96 @@ case class TableJoin(left: TableIR, right: TableIR, joinType: String, joinKey: I
   }
 }
 
+// Possible TODO: Add ability to join on a prefix, rather than just the full keys
+case class TableMultiWayZipJoin(children: IndexedSeq[TableIR], fieldName: String, globalName: String) extends TableIR {
+  require(children.length > 0, "there must be at least one table as an argument")
+
+  private val first = children.head
+  private val rest = children.tail
+
+  require(
+    rest.forall(e => e.typ.keyType isIsomorphicTo first.typ.keyType),
+    "all keys must be the same type"
+  )
+  require(rest.forall(e => e.typ.rowType == first.typ.rowType), "all rows must have the same type")
+  require(rest.forall(e => e.typ.globalType == first.typ.globalType),
+    "all globals must have the same type")
+
+  private val rvdType = RVDType(first.typ.rowType, first.typ.key)
+  private val newGlobalType = TStruct(globalName -> TArray(first.typ.globalType))
+  private val newValueType = TStruct(fieldName -> TArray(rvdType.valueType))
+  private val newRowType = rvdType.kType ++ newValueType
+
+  def typ: TableType = first.typ.copy(
+    rowType = newRowType,
+    globalType = newGlobalType
+  )
+
+  def copy(newChildren: IndexedSeq[BaseIR]): BaseIR =
+    TableMultiWayZipJoin(newChildren.asInstanceOf[IndexedSeq[TableIR]], fieldName, globalName)
+
+
+  def execute(hc: HailContext): TableValue = {
+    val rowType = rvdType.rowType
+    val keyIdx = rvdType.kFieldIdx
+    val valIdx = rvdType.valueFieldIdx
+    val localRVDType = rvdType
+    val localNewRowType = newRowType
+    val rvMerger = { it: Iterator[Array[RegionValue]] =>
+      val rvb = new RegionValueBuilder()
+      val newRegionValue = RegionValue()
+
+      it.map { rvs =>
+        val len = rvs.length
+        val rv = rvs.dropWhile(rv => rv == null).head
```

Comment:
This `dropWhile` is returning a duplicate array.  I don't think there is a Scala-idiomatic way to do this without allocating.  I'd just write a while loop.

Reply:
If we change the return type of the iterator to return `ArrayBuilder[(Int, RV)]`, then we just need to take the first element of that list. Otherwise, I think this should be scala's [`Array.find`](https://www.scala-lang.org/api/2.12.3/scala/Array.html#find(p:A=%3EBoolean):Option[A])